src/logger/src/logger.rs

```diff
@@ -406,71 +323,57 @@ impl Logger {
     /// }
     /// ```
     pub fn init(&self, header: String, log_dest: Box<dyn Write + Send>) -> Result<()> {
-        self.try_lock(INITIALIZING)?;
-        {
-            let mut g = buf_guard(&self.log_buf);
-
-            *g = Some(log_dest);
-        }
+        self.init
+            .call_init(|| {
+                let mut g = extract_guard(self.log_buf.lock());
+                *g = log_dest;
 
-        self.try_init_max_level();
+                // finish init
+                true
+            })
+            .map_err(LoggerError::Init)?;
 
-        STATE.store(INITIALIZED, Ordering::SeqCst);
         self.write_log(header, Level::Info);
 
         Ok(())
     }
 
-    // In a future PR we'll update the way things are written to the selected destination to avoid
-    // the creation and allocation of unnecessary intermediate Strings. The `write_log` method takes
-    // care of the common logic involved in writing regular log messages.
-    fn write_log(&self, msg: String, msg_level: Level) {
-        if STATE.load(Ordering::Relaxed) == INITIALIZED {
-            if let Some(guard) = buf_guard(&self.log_buf).as_mut() {
-                // No need to explicitly call flush because the underlying LineWriter flushes
-                // automatically whenever a newline is detected (and we always end with a
-                // newline the current write).
-                if guard.write_all(&(format!("{}\n", msg)).as_bytes()).is_err() {
-                    // No reason to log the error to stderr here, just increment the metric.
-                    METRICS.logger.missed_log_count.inc();
-                }
-            } else {
-                METRICS.logger.missed_log_count.inc();
-                panic!("Failed to write to the provided log destination due to poisoned lock");
-            }
-        } else if msg_level <= Level::Warn {
-            eprintln!("{}", msg);
+    /// The `write_log` method takes care of the common logic involved in writing
+    /// regular log messages.
+    fn write_log(&self, mut msg: String, msg_level: Level) {
+        let mut guard;
+        let mut dest: Box<dyn Write + Send> = if self.init.is_initialized() {
+            guard = extract_guard(self.log_buf.lock());
+            Box::new(guard.as_mut())
         } else {
-            println!("{}", msg);
+            match msg_level {
```

Comment:
Done.

Reply:
I was referring to adding documentation in the `logger.md` file. I am thinking that this might not be an expected logging behavior, and should be stated in the docs. What do you think?

Reply:
Sure, but that should be already documented:

```
//! 1) Calling `LOGGER.configure()`. This will enable the logger to work in limited mode.
//! In this mode the logger can only write messages to stdout or stderr.
```